app/models/downloaded_file.rb

```diff
@@ -0,0 +1,30 @@
+require 'rmt/checksum_verifier'
+
+class DownloadedFile < ApplicationRecord
+
+  def self.add_file!(checksum_type, checksum, local_path)
+    return nil unless local_path.match?(/\.(rpm|drpm)$/)
+    file_size = File.size(local_path)
+    DownloadedFile.create({ checksum_type: checksum_type,
+                            checksum: checksum,
+                            local_path: local_path,
+                            file_size: file_size })
+  rescue StandardError => e
+    # de-duplication is not critical, just move on
+    logger.error e.message
+    logger.error e.backtrace.each { |line| logger.error line }
+  end
+
+  def self.get_local_path_by_checksum(checksum_type, checksum)
+    file = DownloadedFile.find_by({ checksum_type: checksum_type,
+                                    checksum: checksum })
+    return nil if file.nil? || file.file_size != File.size(file.local_path)
```

Comment:
Testing on FAT filesystem with hardlinks, here I'm getting:

```
No such file or directory @ rb_file_s_size - /home/ivan/work/rmt/public/repo/SUSE/Updates/SLE-SERVER-INSTALLER/12-SP2/x86_64/update/x86_64/hwinfo-21.31_21.38-11.17_15.7.2.x86_64.drpm
/home/ivan/work/rmt/app/models/downloaded_file.rb:21:in `size'
/home/ivan/work/rmt/app/models/downloaded_file.rb:21:in `get_local_path_by_checksum'
```

It might make sense to chech that `file.local_path` exists first.

Reply:
The exception occurs when the file record exists in the DB, but the file has been deleted.